### PR TITLE
Add default constructor for `MapRef`

### LINKED ISSUE #4136: Default constructor for `MapRef`
We currently have many `MapRef` implementations, based on `Ref[Map]`, sharding, `ConcurrentHashMap`, etc. We should have an `apply` constructor that chooses a reasonable default implementation, _probably_ `ConcurrentHashMap`.

### PR DESCRIPTION
**Pull request by yisraelU**
Adds an apply method for MapRef. If the Effect type is an instance of Async , the `ConcurrentHashMap` implementation will be used otherwise the `sharedImmutableMap` will be used with the number of shards set to the number of Processors determined by `Runtime.getRuntime.availableProcessors`.
The goal of this MR is to improve the UX for a user by having a sane default

closes #4136

P.S. as this is a first-time PR to this repo, please let me know if there are any conventions/practices, etc.. that I should be adhering to. TY

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -1,39 +1,5 @@
 {
-  "name": "cats-effect",
```
Comment (BalmungSan) [inline package-lock.json]: I am not sure why there is a `package-lock` in this repo. But, I doubt you wanted to modify it :)
I guess it is best to revert these changes.

File: std/shared/src/main/scala/cats/effect/std/MapRef.scala [hunk 2]
```diff
@@ -45,6 +45,22 @@ trait MapRef[F[_], K, V] extends Function1[K, Ref[F, V]] {
...
+  /**
+   * the default Constructor for MapRef. If Async is available, it will use a ConcurrentHashMap,
+   * otherwise it will use a sharded immutable map.
+   */
+  def apply[F[_]: Concurrent, K, V](): F[MapRef[F, K, Option[V]]] = {
+
+    Concurrent[F] match {
+      case a: Async[_] =>
+        implicit val async: Async[F] = a
+        ofConcurrentHashMap()
```
Comment (BalmungSan) [inline std/shared/src/main/scala/cats/effect/std/MapRef.scala]: IIRC you can match on `Async[F]`, while yes the `F` will be erased it would preserve the right type.

Comment (BalmungSan) [reply std/shared/src/main/scala/cats/effect/std/MapRef.scala]: Nitpick: rather than creating a dummy `implicit val`, you may pass `a` explicitly to `ofConcurrentHashMap()(a)`

File: std/shared/src/main/scala/cats/effect/std/MapRef.scala [hunk 3]
```diff
@@ -45,6 +45,22 @@ trait MapRef[F[_], K, V] extends Function1[K, Ref[F, V]] {
...
+      case _ =>
+        ofShardedImmutableMap[F, K, V](Runtime.getRuntime.availableProcessors())
```
Comment (BalmungSan) [inline std/shared/src/main/scala/cats/effect/std/MapRef.scala]: Nitpick: I would use named arguments and remove the empty parenthesis from `availableProcessors`

File: std/shared/src/main/scala/cats/effect/std/MapRef.scala (hunk 2 above)
Comment (BalmungSan) [reply std/shared/src/main/scala/cats/effect/std/MapRef.scala]: Oh it seems `ofConcurrentHashMap` only requires `Sync`, thus we should only check for that rather than `Async`
Apologies for the confusion, I thought it would require `Async` 😅

File: std/shared/src/main/scala/cats/effect/std/MapRef.scala [hunk 4]
```diff
@@ -45,6 +45,22 @@ trait MapRef[F[_], K, V] extends Function1[K, Ref[F, V]] {
 
 object MapRef extends MapRefCompanionPlatform {
 
+  /**
+   * the default Constructor for MapRef. If Async is available, it will use a ConcurrentHashMap,
+   * otherwise it will use a sharded immutable map.
+   */
+  def apply[F[_]: Concurrent, K, V](): F[MapRef[F, K, Option[V]]] = {
```
Comment (BalmungSan) [inline std/shared/src/main/scala/cats/effect/std/MapRef.scala]: `apply` should not have an empty parameter list.

**Review by BalmungSan — COMMENTED** [review]
While this alone is great.
We may also request a default value in `apply` to improve UX as suggested by @Jasper-M

File: package-lock.json (hunk 1 above)
Comment (armanbilge) [reply package-lock.json]: We need an npm package for testing tracing on JS.

https://github.com/typelevel/cats-effect/blob/eb231ffb0f5368792957a8f955b5836556b8a3e6/package.json#L3-L5

File: std/shared/src/main/scala/cats/effect/std/MapRef.scala (hunk 3 above)
Comment (armanbilge) [reply std/shared/src/main/scala/cats/effect/std/MapRef.scala]: > remove the empty parenthesis from `availableProcessors`

It's technically side-effectual, it may return different values at different times.

- https://github.com/typelevel/cats-effect/issues/3909

Comment (yisraelU) [reply std/shared/src/main/scala/cats/effect/std/MapRef.scala]: @BalmungSan @armanbilge 
So I am confused about this. Should this call be viewed as not referentially transparent and wrapped in IO ? 
I noticed, in other places such as https://github.com/typelevel/cats-effect/blob/eb231ffb0f5368792957a8f955b5836556b8a3e6/core/jvm/src/main/scala/cats/effect/IOApp.scala#L191 directly without being wrapped in IO and its called with the parens 

Also from the JavaDoc 
```This value may change during a particular invocation of the virtual machine. Applications that are sensitive to the number of available processors should therefore occasionally poll this property and adjust their resource usage appropriately.```

File: std/shared/src/main/scala/cats/effect/std/MapRef.scala (hunk 2 above)
Comment (yisraelU) [reply std/shared/src/main/scala/cats/effect/std/MapRef.scala]: > Nitpick: rather than creating a dummy `implicit val`, you may pass `a` explicitly to `ofConcurrentHashMap()(a)`

Cool , Ididn't know you could do that when a method is defined with a Context Bound

File: package-lock.json (hunk 1 above)
Comment (yisraelU) [reply package-lock.json]: It's coming from something autogenerated . Ill try to see how this occured

Comment (yisraelU) [reply package-lock.json]: fixed

**Comment by yisraelU** [comment]
> While this alone is great. Wwe may also request a default value in `apply` to improve UX as suggested by @Jasper-M

I have added an overloaded apply . WDYT

File: std/shared/src/main/scala/cats/effect/std/MapRef.scala (hunk 3 above)
Comment (armanbilge) [reply std/shared/src/main/scala/cats/effect/std/MapRef.scala]: > So I am confused about this. Should this call be viewed as not referentially transparent and wrapped in IO ?

Technically, you are completely right. But we have decided to cheat about this in several places as you've noticed, because it would be impractical to properly suspend this.

See also https://github.com/typelevel/cats-effect/issues/3909. tl;dr we are assuming it's a static value.

Comment (BalmungSan) [reply std/shared/src/main/scala/cats/effect/std/MapRef.scala]: Right my bad, I thought that value may not change.
Yeah, is best to leave the parenthesis then. Ideally, it should even be wrapped but here we can't wrap it so as you have it is the best :)

**Comment by BalmungSan** [comment]
> I have added an overloaded apply . WDYT

Sadly I don't think that would work well: https://scastie.scala-lang.org/BalmungSan/WHoq2ShkTCm3QSs5puHixw

I think the best would be to do what Arman suggested and rather add a `withDefault` _(better names are welcome)_ extension method to `MapRefOptionOps`. So that users can do:

```scala
val m1 = MapRef[IO, K, V] // Returns MapRef[IO, K, Option[V]
val m2 = m1.withDefault(foo) // Returns MapRef[IO, K, V]
```

**Comment by yisraelU** [comment]
> > I have added an overloaded apply . WDYT
> 
> Sadly I don't think that would work well: https://scastie.scala-lang.org/BalmungSan/WHoq2ShkTCm3QSs5puHixw
> 
> I think the best would be to do what Arman suggested and rather add a `withDefault` _(better names are welcome)_ extension method to `MapRefOptionOps`. So that users can do:
> 
> ```scala
> val m1 = MapRef[IO, K, V] // Returns MapRef[IO, K, Option[V]
> val m2 = m1.withDefault(foo) // Returns MapRef[IO, K, V]
> ```

I see, it fails for certain types.
If we go back to empty parens for the apply method it should work. 
I'm happy with the extension method solution though, and I have nothing better than `withDefault` :)

**Review by BalmungSan — APPROVED** [review]

File: std/shared/src/main/scala/cats/effect/std/MapRef.scala [hunk 5]
```diff
@@ -45,6 +45,20 @@ trait MapRef[F[_], K, V] extends Function1[K, Ref[F, V]] {
...
+  /**
+   * the default Constructor for MapRef. If Async is available, it will use a ConcurrentHashMap,
+   * otherwise it will use a sharded immutable map.
+   */
+  def apply[F[_]: Concurrent, K, V]: F[MapRef[F, K, Option[V]]] = {
+    Concurrent[F] match {
+      case s: Sync[F] =>
+        ofConcurrentHashMap()(s)
+      case _ =>
+        ofShardedImmutableMap[F, K, V](shardCount = Runtime.getRuntime.availableProcessors())
```
Comment (durban) [inline std/shared/src/main/scala/cats/effect/std/MapRef.scala:57]: The `availableProcessors` call is technically side-effecting, so should be inside `F`. But here we don't have `Sync`, so I don't really know what to do...

Comment (armanbilge) [reply std/shared/src/main/scala/cats/effect/std/MapRef.scala:57]: See https://github.com/typelevel/cats-effect/pull/4139#discussion_r1763988320

File: std/shared/src/main/scala/cats/effect/std/MapRef.scala [hunk 6]
```diff
@@ -45,6 +45,20 @@ trait MapRef[F[_], K, V] extends Function1[K, Ref[F, V]] {
 
 object MapRef extends MapRefCompanionPlatform {
 
+  /**
+   * the default Constructor for MapRef. If Async is available, it will use a ConcurrentHashMap,
+   * otherwise it will use a sharded immutable map.
+   */
```
Comment (armanbilge) [inline std/shared/src/main/scala/cats/effect/std/MapRef.scala]:
```suggestion
  /**
   * Default constructor for [[MapRef]]. If [[Sync]] is available, it will delegate to [[ofConcurrentHashMap],
   * otherwise it will fallback to [[ofShardedImmutableMap]].
   */
```

Comment (yisraelU) [reply std/shared/src/main/scala/cats/effect/std/MapRef.scala]: Thanks . I forgot to update after changing to Sync

File: std/shared/src/main/scala/cats/effect/std/MapRef.scala (hunk 5 above)
Comment (durban) [reply std/shared/src/main/scala/cats/effect/std/MapRef.scala:57]: Thanks, sorry I missed that.

**Review by BalmungSan — APPROVED** [review]

File: std/shared/src/main/scala/cats/effect/std/MapRef.scala [hunk 7]
```diff
@@ -45,6 +45,20 @@ trait MapRef[F[_], K, V] extends Function1[K, Ref[F, V]] {
 
 object MapRef extends MapRefCompanionPlatform {
 
+  /**
+   * Default constructor for [[MapRef]]. If [[Sync]] is available, it will delegate to
+   * [[ofConcurrentHashMap], otherwise it will fallback to [[ofShardedImmutableMap]].
```
Comment (armanbilge) [inline std/shared/src/main/scala/cats/effect/std/MapRef.scala]:
```suggestion
   * [[ofConcurrentHashMap]], otherwise it will fallback to [[ofShardedImmutableMap]].
```

**Review by durban — APPROVED** [review]

**Review by armanbilge — APPROVED** [review]
thanks!
